### PR TITLE
CASSANDRA-18124 Make keystore_password nullable

### PR DESCRIPTION
**Pull request by maulin-vasavada**
The JIRA Ticket: [CASSANDRA-18124 JIRA](https://issues.apache.org/jira/browse/CASSANDRA-18124)

```
CASSANDRA-18124 Make keystore_password nullable.

Changes made:
1. Fixed the compilation issue for examples/ssl-factory module on trunk
2. Made keystore_password/outbound_keystore_password and truststore_password configurations nullable
3. Added keystore_password/outbound_keystore_password and truststore_password required for File based sslcontextfactories (except PEM)
4. Removed the warning upon keystore_password and key_password both are provided but match the values
5. Added a reference to the documentation for configuring the PEM based key material in the cassandra.yaml

```

Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>
```

### CODE REVIEW

File: src/java/org/apache/cassandra/security/FileBasedSslContextFactory.java
Comment (smiklosovic) [inline src/java/org/apache/cassandra/security/FileBasedSslContextFactory.java]: @maulin-vasavada you are using illegal imports, you should use `org.apache.cassandra.io.util.File.getPath` instead

````
checkstyle:
    [mkdir] Created dir: /home/fermat/dev/cassandra/cassandra-instaclustr/cassandra/build/checkstyle
[checkstyle] Running Checkstyle 8.45.1 on 2201 files
[checkstyle] [ERROR] /home/fermat/dev/cassandra/cassandra-instaclustr/cassandra/src/java/org/apache/cassandra/security/FileBasedSslContextFactory.java:23:1: Illegal import - java.nio.file.Paths. [IllegalImport]
````

Comment (maulin-vasavada) [reply src/java/org/apache/cassandra/security/FileBasedSslContextFactory.java]: Sure, made the change. Thanks. Will figure out why I am unable to see the same checkstyle error locally.

File: conf/cassandra.yaml
Comment (smiklosovic) [inline conf/cassandra.yaml]: I am not completely sure about this URL. There is no case of us adding links like this in whole cassandra.yaml.

Comment (maulin-vasavada) [reply conf/cassandra.yaml]: One of the ask on the ticket was to include an example in comments for the PEM formatted keys. However, since we have client and server both options, adding long repeated comments may not make sense. Hence I added a link like another http link I saw on the yaml document. Would love to hear from you and other members on this and get more direction.

Comment (maedhroz) [reply conf/cassandra.yaml]: There is something similar for `num_tokens`, but even that is sort of a "best practices" reference, not something that describes the configuration format itself. The problem here is that what is at the URL in question won't necessarily be versioned along w/ the code itself, right?

We already duplicate a small amount of information in the comments for server and client encryption. What about just adding a commented out section before `keystore` to both that looks something like this:

```
# Configure the way Cassandra creates SSL contexts.
# To use PEM-based key material, see org.apache.cassandra.security.PEMBasedSslContextFactory
# ssl_context_factory:
#     # Must be an instance of org.apache.cassandra.security.ISslContextFactory
#     class_name: org.apache.cassandra.security.DefaultSslContextFactory
```

`PEMBasedSslContextFactory` at least has comments that should be updated if necessary.

(As an aside, this is one of the reasons something like CASSANDRA-17292 would be useful. There's no documentation-friendly hierarchy in our configuration format.)

File: src/java/org/apache/cassandra/config/EncryptionOptions.java
Comment (maedhroz) [inline src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: What's the worst thing that could happen as a result of this change? A `cassandra.yaml` out in the wild that uses "cassandra" as a password implicitly by virtue of not specifying it will suddenly break unless the password is explicitly set to the old default?

That's probably fine in a major release, in any case...

**Review by maedhroz — APPROVED** [review]
Approved, w/ some minor commentary.

File: src/java/org/apache/cassandra/config/EncryptionOptions.java
Comment (smiklosovic) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: @maedhroz 

I do not think that having passwords defaulting to some value is a good idea anyway. A password should be of some complex form, not "cassandra". What value is in that, anyway? It would be quite weird to have a commented password field and then having a certificate / keystore / truststore to have "cassandra" password. That is just strange and I dont think that is used in practice a lot ....

File: conf/cassandra.yaml
Comment (smiklosovic) [reply conf/cassandra.yaml]: I will leave this to @maulin-vasavada to decide.

File: src/java/org/apache/cassandra/config/EncryptionOptions.java
Comment (rtib) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: While I understand the concern about changing a default value belongs to a major release, I'm not sure that the original issue should wait for 5.0. Any solution shipped with the next 4.1 release is considered better than postpone to 5.0.

Would it possible to leave the default value `cassandra` in place while enabling it to be set null explicitly? IMO, that wouldn't conflict with SemVer policy and could be release with the next 4.1.

Comment (smiklosovic) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: @maulin-vasavada what do you think about @rtib comment above? I think that would be the most ideal solution which might be shipped in next 4.1 patch release (probably in 4.1.2).

Comment (maedhroz) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: Either way, I would probably add an entry in `NEWS.txt` for 5.0 in the "Upgrading" section to explain the default change.

> I do not think that having passwords defaulting to some value is a good idea anyway...That is just strange and I dont think that is used in practice a lot

100% agree, and I don't know why we ever did it that way in the first place :-|

File: conf/cassandra.yaml
Comment (maulin-vasavada) [reply conf/cassandra.yaml]: Thanks @maedhroz for the input. I like that idea. Changed the documentation like suggested. Please review.

File: src/java/org/apache/cassandra/config/EncryptionOptions.java
Comment (maulin-vasavada) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: If we keep the default value in the EncryptionOptions then it becomes difficult to have tests working for PEM use-cases for the unencrypted key. Hence I would like to keep the null assignment in the code with these changes. If that bumps this upto go in 5.x I am fine with it.

File: conf/cassandra.yaml
Comment (maedhroz) [inline conf/cassandra.yaml:1419]: Looks good...not too much duplication in the end.

File: src/java/org/apache/cassandra/config/EncryptionOptions.java
Comment (maedhroz) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: I'm fine either way as long as `NEWS.txt` mentions it.

Comment (maulin-vasavada) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: @maedhroz Do we need the change in NEWS.txt in this PR or that would be part of the 5.0 release? I can resolve this conversation after this for a committer to merge this.

Comment (maedhroz) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: Yes, even if it's in a major release, a change in behavior should be documented there.

Comment (smiklosovic) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: I am for documenting it and merge it to 4.1 as well as to trunk. I think this patch is for trunk only, currently. I can backport it to 4.1 quite easily, should be nicely backportable.

Comment (maulin-vasavada) [reply src/java/org/apache/cassandra/config/EncryptionOptions.java:156]: @maedhroz and @smiklosovic I added a text in the News txt file under `Upgrades` for 5.0 for the removal of the default keystore passwords. Please take a look. After that we are good to merge this.

File: NEWS.txt
Comment (maedhroz) [inline NEWS.txt]:
```suggestion
     removed. Any deployments that depend upon the code default to this password value without explicitly specifying
```

File: NEWS.txt
Comment (maedhroz) [inline NEWS.txt]:
```suggestion
     it in cassandra.yaml will fail on upgrade. Please specify your keystore_password and truststore_password elements in cassandra.yaml with appropriate
```

File: NEWS.txt
Comment (maedhroz) [inline NEWS.txt]:
```suggestion
     values to prevent this failure.
```

File: conf/cassandra.yaml
Comment (maedhroz) [inline conf/cassandra.yaml]: Looking at this one last time, don't we want to comment out all 4 of the `keystore_password` and `truststore_password` field in the YAML? If we leave them there, and operators just start using the YAML we ship, we will essentially have preserved the "cassandra" default. The point of all this is that we want to force the operator to choose passwords, right?

CC @smiklosovic

Comment (maulin-vasavada) [reply conf/cassandra.yaml]: I agree with you @maedhroz. Will wait to hear Stefan's comment and would make appropriate changes.

Comment (maulin-vasavada) [reply conf/cassandra.yaml]: @smiklosovic any comments before we can close this?

Comment (smiklosovic) [reply conf/cassandra.yaml]: @maedhroz  @maulin-vasavada 

But if we leave it here uncommented, we also do not break existing configurations which are depending on it.

It is about "not breaking what is there". If we change this, all deployments which rely on the default will stop to work because we set it to null suddenly.

It is more about the _possibility_ to set it to null if one really wants that rather than setting it by default to null and break it everywhere it relies on not null (even it is "cassandra" as default).

People who are (and should) using stronger passwords have to set them explicitly anyway and they do not have this problem.

If this goes to 4.1 as well, people might be quite surprised that this will not work between stupid version bump in a patch release. Having it changed in 5.0 is probably OK but still ... 

So, I would leave it uncommented in 4.1 but I would comment it in 5.0.

That way we do not break patch release bump but we still educate users that there is no magic default for them in 5.0 and they should set it all explicitly.

File: NEWS.txt
Comment (smiklosovic) [inline NEWS.txt]: @maulin-vasavada we should mention keystore as well as truststore passwords here. It would be also nice to mention what section these configuration parameters are found in.

File: conf/cassandra.yaml
Comment (maedhroz) [reply conf/cassandra.yaml]: That works for me.

Comment (maulin-vasavada) [reply conf/cassandra.yaml]: Cool. So @smiklosovic and @maedhroz I have [a separate PR](https://github.com/apache/cassandra/pull/2273) for the 4.1 backport and in this PR I can comment out the password configuration from the cassandra yaml since this will be target for 5.0 release. Please let me know if that understanding is correct.

Comment (maulin-vasavada) [reply conf/cassandra.yaml]: Done commenting out the passwords in the cassandra yaml in this PR.

File: NEWS.txt
Comment (maulin-vasavada) [reply NEWS.txt]: done

File: conf/cassandra.yaml
Comment (smiklosovic) [reply conf/cassandra.yaml]: thank you, I will take a look shortly. Thanks for the patience.

Comment (smiklosovic) [reply conf/cassandra.yaml]: @maulin-vasavada looks good, please apply what was discussed here in 4.1 branch and we can finally ship it.

EDIT: Actually, we do not need to do anything for 4.1 in NEWS because we effectively have not changed anything. Upgrades will be working as before.

So I think we are done.

Comment (maulin-vasavada) [reply conf/cassandra.yaml]: Cool thanks @smiklosovic I've the backported changes for 4.1 [here in a different PR](https://github.com/apache/cassandra/pull/2273). Will remove news.txt update from there. Thanks for helping.

**Pull request closed without merging**
